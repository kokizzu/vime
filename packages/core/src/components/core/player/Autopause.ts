let players: HTMLVimePlayerElement[] = [];

export class Autopause {
  constructor(private readonly player: HTMLVimePlayerElement) {
    players.push(player);
  }

  willPlay() {
    players.forEach((p) => {
      try {
        // eslint-disable-next-line no-param-reassign
        if (p !== this.player && p.autopause) p.paused = true;
      } catch (e) {
        // Might throw when testing because `disconnectCallback` isn't called.
      }
    });
  }

  destroy() {
    players = players.filter((p) => p !== this.player);
  }
}
