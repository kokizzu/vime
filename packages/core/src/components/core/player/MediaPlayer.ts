import { ComponentInterface } from '@stencil/core';
import { PlayerProps } from './PlayerProp';
import { PlayerEvents } from './PlayerEvent';
import { PlayerMethods } from './PlayerMethod';

/**
 * The core media player interface.
 *
 * @ref https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement
 */
export interface MediaPlayer extends ComponentInterface, PlayerProps, PlayerEvents, PlayerMethods {}
