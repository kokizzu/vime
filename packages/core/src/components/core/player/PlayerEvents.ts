import { EventEmitter } from '@stencil/core';
import { PlayerProp, PlayerProps } from './PlayerProps';

// Events that toggle state and the prop is named `is{PropName}...`.
const isToggleStateEvent = new Set<PlayerProp>([
  'isFullscreenActive',
  'isControlsActive',
  'isPiPActive',
  'isLive',
  'isTouch',
  'isAudio',
  'isVideo',
  'isAudioView',
  'isVideoView',
]);

// Events that are emitted without the 'Change' postfix.
const hasShortenedEventName = new Set<PlayerProp>([
  'ready',
  'playbackStarted',
  'playbackEnded',
  'playbackReady',
]);

export const getEventName = (prop: PlayerProp): keyof PlayerEvents => {
  // Example: isFullscreenActive -> vFullscreenChange
  if (isToggleStateEvent.has(prop)) {
    return `v${prop.replace('is', '').replace('Active', '')}Change` as any;
  }

  // Example: playbackStarted -> vPlaybackStarted
  if (hasShortenedEventName.has(prop)) {
    return `v${prop.charAt(0).toUpperCase()}${prop.slice(1)}` as any;
  }

  // Example: currentTime -> vCurrentTimeChange
  return `v${prop.charAt(0).toUpperCase()}${prop.slice(1)}Change` as any;
};

export type PlayerEvent = keyof PlayerEvents;

export interface PlayerEvents {
  /**
   * Emitted when the player is attached/deattached from the DOM.
   */
  vAttachedChange: EventEmitter<void>;

  /**
   * Emitted when the `theme` prop changes value.
   */
  vThemeChange: EventEmitter<PlayerProps['theme']>;

  /**
   * Emitted when the `paused` prop changes value.
   */
  vPausedChange: EventEmitter<PlayerProps['paused']>

  /**
   * Emitted when the media is transitioning from `paused` to `playing`. Event flow: `paused` ->
   * `play` -> `playing`. The media starts `playing` once enough content has buffered to
   * begin/resume playback.
   */
  vPlay: EventEmitter<void>

  /**
   * Emitted when the `playing` prop changes value.
   */
  vPlayingChange: EventEmitter<PlayerProps['playing']>

  /**
   * Emitted when the `seeking` prop changes value.
   */
  vSeekingChange: EventEmitter<PlayerProps['seeking']>

  /**
   * Emitted directly after the player has successfully transitioned/seeked to a new time position.
   * Event flow: `seeking` -> `seeked`.
   */
  vSeeked: EventEmitter<void>

  /**
   * Emitted when the `buffering` prop changes value.
   */
  vBufferingChange: EventEmitter<PlayerProps['buffering']>

  /**
   * Emitted when the `duration` prop changes value.
   */
  vDurationChange: EventEmitter<PlayerProps['duration']>

  /**
   * Emitted when the `currentTime` prop changes value.
   */
  vCurrentTimeChange: EventEmitter<PlayerProps['currentTime']>

  /**
   * Emitted when the player has loaded and is ready to be interacted with.
   */
  vReady: EventEmitter<void>;

  /**
   * Emitted when the media is ready to begin playback. The following props are guaranteed to be
   * defined when this fires: `mediaTitle`, `currentSrc`, `currentPoster`, `duration`, `mediaType`,
   * `viewType`.
   */
  vPlaybackReady: EventEmitter<void>;

  /**
   * Emitted when the media initiates playback.
   */
  vPlaybackStarted: EventEmitter<void>

  /**
   * Emitted when playback reaches the end of the media.
   */
  vPlaybackEnded: EventEmitter<void>

  /**
   * Emitted when the `buffered` prop changes value.
   */
  vBufferedChange: EventEmitter<PlayerProps['buffered']>

  /**
   * Emitted when the `currentSrc` prop changes value.
   */
  vCurrentSrcChange: EventEmitter<PlayerProps['currentSrc']>

  /**
   * Emitted when the `currentPoster` prop changes value.
   */
  vCurrentPosterChange: EventEmitter<PlayerProps['currentPoster']>

  /**
   * Emitted when the `mediaTitle` prop changes value.
   */
  vMediaTitleChange: EventEmitter<PlayerProps['mediaTitle']>

  /**
   * Emitted when the `errors` prop changes value.
   */
  vErrorsChange: EventEmitter<PlayerProps['errors']>

  /**
   * Emitted when the `textTracks` prop changes value.
   */
  vTextTracksChange: EventEmitter<PlayerProps['textTracks']>

  /**
   * Emitted when the provider starts loading a media resource.
   */
  vLoadStart: EventEmitter<void>

  /**
   * Emitted when the `playbackRate` prop changes value.
   */
  vPlaybackRateChange: EventEmitter<PlayerProps['playbackRate']>

  /**
   * Emitted when the `playbackRates` prop changes value.
   */
  vPlaybackRatesChange: EventEmitter<PlayerProps['playbackRates']>

  /**
   *
   * Emitted when the `playbackQuality` prop changes value.
   */
  vPlaybackQualityChange: EventEmitter<PlayerProps['playbackQuality']>

  /**
   * Emitted when the `playbackQualities` prop changes value.
   */
  vPlaybackQualitiesChange: EventEmitter<PlayerProps['playbackQualities']>

  /**
   * Emitted when the `muted` prop changes value.
   */
  vMutedChange: EventEmitter<PlayerProps['muted']>

  /**
   * Emitted when the `volume` prop changes value.
   */
  vVolumeChange: EventEmitter<PlayerProps['volume']>

  /**
   * Emitted when the `mediaType` prop changes value.
   */
  vMediaTypeChange: EventEmitter<PlayerProps['mediaType']>

  /**
   * Emitted when the `viewType` prop changes value.
   */
  vViewTypeChange: EventEmitter<PlayerProps['viewType']>

  /**
   * Emitted when the `isControlsActive` prop changes value.
   */
  vControlsChange: EventEmitter<PlayerProps['isControlsActive']>

  /**
   * Emitted when the `currentCaption` prop changes value.
   */
  vCurrentCaptionChange: EventEmitter<PlayerProps['currentCaption']>

  /**
   * Emitted when the `isLive` prop changes value.
   */
  vLiveChange: EventEmitter<PlayerProps['isLive']>

  /**
   * Emitted when the `isTouch` prop changes value.
   */
  vTouchChange: EventEmitter<PlayerProps['isTouch']>

  /**
   * Emitted when the `language` prop changes value.
   */
  vLanguageChange: EventEmitter<PlayerProps['language']>

  /**
   * Emitted when the `languages` prop changes value.
   */
  vLanguagesChange: EventEmitter<PlayerProps['languages']>

  /**
   * Emitted when the `i18n` prop changes value.
   */
  vI18nChange: EventEmitter<PlayerProps['i18n']>

  /**
   * Emitted when the `translations` prop changes value.
   */
  vTranslationsChange: EventEmitter<PlayerProps['translations']>

  /**
   * Emitted when the `isFullscreenActive` prop changes value.
   */
  vFullscreenChange: EventEmitter<PlayerProps['isFullscreenActive']>

  /**
   * Emitted when the `isPiPActive` prop changes value.
   */
  vPiPChange: EventEmitter<PlayerProps['isPiPActive']>
}
