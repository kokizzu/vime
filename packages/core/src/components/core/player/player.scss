vime-player {
  /**
   * @prop --vime-font-family: A custom font family to be used throughout the player.
   * 
   * @prop --vime-theme: A custom theme (color) to be used throughout the player. Any valid CSS 
   * `color` property should work (HEX, RGBA, HLS, ...).
   */

  --vime-font-family: 'Helvetica Neue', 'Segoe UI', Helvetica, Arial, sans-serif;

  box-sizing: border-box;
  direction: ltr;
  font-family: var(--vime-font-family);
  -moz-osx-font-smoothing: auto;
  -webkit-font-smoothing: subpixel-antialiased;
  -webkit-tap-highlight-color: transparent;
  font-variant-numeric: tabular-nums;
  font-weight: 500;
  line-height: 1.7;
  width: 100%;
  display: block;
  max-width: 100%;
  min-width: 414px;
  position: relative;
  text-shadow: none;
  outline: 0;
  transition: box-shadow 0.3s ease;

  *,
  *::after,
  *::before {
    box-sizing: inherit;
  }

  button {
    font: inherit;
    line-height: inherit;
    width: auto;

    &::-moz-focus-inner {
      border:0;
    }
  }

  // Fix 300ms delay
  a,
  button,
  input,
  label {
    touch-action: manipulation;
  }

  &.video {
    height: 0;
    background: #000;
    overflow: hidden;
  }

  &.fullscreen {
    background: #000;
    margin: 0;
    border-radius: 0;
    width: 100%;
    height: 100%;
    padding-bottom: 0 !important;
  }

  .blocker {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: $z-index-blocker;
    display: inline-block;
  }
}
