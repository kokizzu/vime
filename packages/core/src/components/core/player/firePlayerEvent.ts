import { PlayerProp } from './PlayerProps';
import { MediaPlayer } from './MediaPlayer';

// Events that toggle state and the prop is named `is{PropName}Active`.
const isToggleStateEvent = new Set([
  PlayerProp.IsFullscreenActive,
  PlayerProp.isPiPActive,
  PlayerProp.IsLive,
  PlayerProp.IsTouch,
]);

// Events that are emitted without the 'Change' postfix.
const hasShortenedEventName = new Set([
  PlayerProp.PlaybackStarted,
  PlayerProp.PlaybackEnded,
  PlayerProp.PlaybackReady,
  PlayerProp.LoadedMetadata,
]);

let startedSeeking = false;

export const firePlayerEvent = (player: MediaPlayer, prop: PlayerProp, value: any) => {
  let eventName;

  if (isToggleStateEvent.has(prop)) {
    // Example: isFullscreenActive -> vFullscreenChange
    eventName = `v${prop.replace('is', '').replace('Active', '')}Change`;
  } else if (hasShortenedEventName.has(prop)) {
    // Example: playbackStarted -> vPlaybackStarted
    eventName = `v${prop.charAt(0).toUpperCase()}${prop.slice(1)}`;
  } else {
    // Example: currentTime -> vCurrentTimeChange
    eventName = `v${prop.charAt(0).toUpperCase()}${prop.slice(1)}Change`;
  }

  player[eventName]?.emit(value);

  if ((prop === PlayerProp.Paused) && !value) player.vPlay.emit();

  if (prop === PlayerProp.Seeking) {
    if (startedSeeking && !value) player.vSeeked.emit();
    startedSeeking = value;
  }
};
