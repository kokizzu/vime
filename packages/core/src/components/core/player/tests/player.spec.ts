import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { NewSpecPageOptions } from '@stencil/core/internal';
import { Player } from '../player';
import { FakeTube } from '../../../providers/faketube/FakeTube';
import { MediaType } from '../MediaType';
import { ViewType } from '../ViewType';
import { MockMediaProviderAdapter } from '../../../providers/MediaProvider';
import { initialState, PlayerProp } from '../PlayerProps';
import { getEventName } from '../PlayerEvents';

let page: SpecPage;
let player: HTMLVimePlayerElement;
let adapter: MockMediaProviderAdapter;
let provider: HTMLVimeFaketubeElement;

const buildPage = async (opts?: Partial<NewSpecPageOptions>) => {
  page = await newSpecPage({
    components: [Player, FakeTube],
    html: '<vime-player debug><vime-faketube /></vime-player>',
    ...opts,
  });

  player = page.root! as HTMLVimePlayerElement;
  provider = page.root!.querySelector('vime-faketube')!;
  adapter = await player.getAdapter() as MockMediaProviderAdapter;
};

beforeEach(() => buildPage());

it('should be structurally sound', async () => {
  expect(page.root).toMatchSnapshot();
});

describe('props', () => {
  it('should have defined all props', async () => {
    const isUndefinedProp = new Set([
      'playbackQuality',
      'theme',
      'mediaType',
      'viewType',
      'mediaTitle',
      'textTracks',
      'currentSrc',
      'currentPoster',
      'currentCaption',
    ]);

    Object.keys(initialState).forEach((prop) => {
      if (!isUndefinedProp.has(prop)) expect((player as any)[prop]).toBeDefined();
    });
  });

  it('should not throw if attempting to change a writable property', async () => {
    player.controls = true;
    await page.waitForChanges();
  });

  it('should watch `mediaType` prop', async () => {
    expect(player.isAudio).toBeFalsy();
    expect(player.isVideo).toBeFalsy();
    await provider.dispatchChange('mediaType', MediaType.Audio);
    await page.waitForChanges();
    expect(player.isAudio).toBeTruthy();
    expect(player.isVideo).toBeFalsy();
    await provider.dispatchChange('mediaType', MediaType.Video);
    await page.waitForChanges();
    expect(player.isAudio).toBeFalsy();
    expect(player.isVideo).toBeTruthy();
  });

  it('should watch `viewType` prop', async () => {
    expect(player.isAudioView).toBeFalsy();
    expect(player.isVideoView).toBeFalsy();
    await provider.dispatchChange('viewType', ViewType.Audio);
    await page.waitForChanges();
    expect(player.isAudioView).toBeTruthy();
    expect(player.isVideoView).toBeFalsy();
    await provider.dispatchChange('viewType', ViewType.Video);
    await page.waitForChanges();
    expect(player.isAudioView).toBeFalsy();
    expect(player.isVideoView).toBeTruthy();
  });

  it('should set playing to false when paused changes to true', async () => {
    player.paused = false;
    await provider.dispatchChange('playing', true);
    await page.waitForChanges();
    player.paused = true;
    await page.waitForChanges();
    expect(player.playing).toBeFalsy();
  });

  it('should update languages prop when a new language is added', async () => {
    await player.extendLanguage('tr', { Pause: 'Duraklat' });
    await page.waitForChanges();
    expect(player.languages).toEqual(['en', 'tr']);
  });

  it('should update i18n when the language or translations are changed', async () => {
    await player.extendLanguage('tr', { Pause: 'Duraklat' });
    player.language = 'tr';
    await page.waitForChanges();
    expect(player.i18n).toEqual({ Pause: 'Duraklat' });
    await player.extendLanguage('tr', { Play: 'Oyna', Pause: 'Duraklat' });
    await page.waitForChanges();
    expect(player.i18n).toEqual({ Play: 'Oyna', Pause: 'Duraklat' });
  });

  it('should call adapter when paused is changed', async () => {
    expect(adapter.play).not.toHaveBeenCalled();
    expect(adapter.pause).not.toHaveBeenCalled();
    await provider.dispatchChange('playbackReady', true);
    player.paused = false;
    await page.waitForChanges();
    await provider.dispatchChange('paused', false);
    await page.waitForChanges();
    expect(adapter.play).toHaveBeenCalled();
    expect(adapter.pause).not.toHaveBeenCalled();
    player.paused = true;
    await page.waitForChanges();
    expect(adapter.play).toHaveBeenCalledTimes(1);
    expect(adapter.pause).toHaveBeenCalled();
  });

  it('should call adapter when muted is changed', async () => {
    expect(adapter.setMuted).not.toHaveBeenCalled();
    await provider.dispatchChange('playbackReady', true);
    player.muted = true;
    await page.waitForChanges();
    expect(adapter.setMuted).toHaveBeenCalledWith(true);
    await provider.dispatchChange('muted', true);
    await page.waitForChanges();
    player.muted = false;
    await page.waitForChanges();
    expect(adapter.setMuted).toHaveBeenCalledWith(false);
  });

  it('should call adapter when volume is changed', async () => {
    expect(adapter.setVolume).not.toHaveBeenCalled();
    await provider.dispatchChange('playbackReady', true);
    player.volume = 30;
    await page.waitForChanges();
    expect(adapter.setVolume).toHaveBeenCalledWith(30);
  });

  it('should call adapter when currentTime is changed', async () => {
    expect(adapter.setCurrentTime).not.toHaveBeenCalled();
    await provider.dispatchChange('duration', 100);
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    player.currentTime = 30;
    await page.waitForChanges();
    expect(adapter.setCurrentTime).toHaveBeenCalledWith(30);
  });

  it('should not change playbackRate if provider can\'t set it', async () => {
    const spy = jest.spyOn(console, 'log').mockImplementation();
    adapter.canSetPlaybackRate!.mockImplementation(() => Promise.resolve(false));
    await provider.dispatchChange('playbackReady', true);
    player.playbackRate = 2;
    await page.waitForChanges();
    expect(player.playbackRate).toEqual(1);
    expect(spy).toHaveBeenCalledWith('[Vime tip]:', expect.stringContaining('change `playbackRate`'));
    spy.mockRestore();
  });

  it('should not change playbackRate if not in playbackRates', async () => {
    const spy = jest.spyOn(console, 'log').mockImplementation();
    adapter.canSetPlaybackRate!.mockImplementationOnce(() => Promise.resolve(true));
    await provider.dispatchChange('playbackReady', true);
    await provider.dispatchChange('playbackRates', [1, 2, 3]);
    await page.waitForChanges();
    player.playbackRate = 4;
    await page.waitForChanges();
    expect(player.playbackRate).toEqual(1);
    expect(spy).toHaveBeenCalledWith('[Vime tip]:', expect.stringContaining('invalid `playbackRate`'));
    spy.mockRestore();
  });

  it('should call adapter when playbackRate is changed', async () => {
    adapter.canSetPlaybackRate!.mockImplementationOnce(() => Promise.resolve(true));
    await provider.dispatchChange('playbackReady', true);
    await provider.dispatchChange('playbackRates', [1, 2]);
    await page.waitForChanges();
    player.playbackRate = 2;
    await page.waitForChanges();
    expect(adapter.setPlaybackRate).toHaveBeenCalledWith(2);
  });

  it('should not change playbackQuality if provider can\'t set it', async () => {
    const spy = jest.spyOn(console, 'log').mockImplementation();
    adapter.canSetPlaybackQuality!.mockImplementationOnce(() => Promise.resolve(false));
    await provider.dispatchChange('playbackReady', true);
    player.playbackQuality = '720p';
    await page.waitForChanges();
    expect(player.playbackQuality).toBeUndefined();
    expect(spy).toHaveBeenCalledWith('[Vime tip]:', expect.stringContaining('change `playbackQuality`'));
    spy.mockRestore();
  });

  it('should not change playbackQuality if not in playbackQualities', async () => {
    const spy = jest.spyOn(console, 'log').mockImplementation();
    adapter.canSetPlaybackQuality!.mockImplementationOnce(() => Promise.resolve(true));
    await provider.dispatchChange('playbackReady', true);
    await provider.dispatchChange('playbackQualities', ['1080p', '720p']);
    await page.waitForChanges();
    player.playbackQuality = '480p';
    await page.waitForChanges();
    expect(player.playbackQuality).toBeUndefined();
    expect(spy).toHaveBeenCalledWith('[Vime tip]:', expect.stringContaining('invalid `playbackQuality`'));
    spy.mockRestore();
  });

  it('should call adapter when playbackQuality is changed', async () => {
    adapter.canSetPlaybackQuality!.mockImplementationOnce(() => Promise.resolve(true));
    await provider.dispatchChange('playbackReady', true);
    await provider.dispatchChange('playbackQualities', ['1080p', '720p']);
    await page.waitForChanges();
    player.playbackQuality = '720p';
    await page.waitForChanges();
    expect(adapter.setPlaybackQuality).toHaveBeenCalledWith('720p');
  });

  // Test doesn't work for some unknown reason.
  it.skip('should throw if writing to internal readonly prop', async () => {
    await expect(provider.dispatchChange('isAudio', true)).rejects.toThrow();
  });

  it('should adjust paddingBottom according to aspectRatio', async () => {
    expect(page.root!.style.paddingBottom).toEqual('');
    await provider.dispatchChange('viewType', ViewType.Video);
    await page.waitForChanges();
    expect(page.root!.style.paddingBottom).toEqual('56.25%');
    await provider.dispatchChange('viewType', ViewType.Audio);
    await page.waitForChanges();
    expect(page.root!.style.paddingBottom).toEqual('');
  });

  it('should render blocker', async () => {
    expect(page.root!.querySelector('.blocker')).toBeNull();
    await provider.dispatchChange('viewType', ViewType.Video);
    await page.waitForChanges();
    expect(page.root!.querySelector('.blocker')).toBeDefined();
  });

  it('should not reset props on first media change', async () => {
    player.currentTime = 40;
    player.volume = 80;
    await page.waitForChanges();
    await provider.dispatchLoadStart();
    await page.waitForChanges();
    expect(player.currentTime).toEqual(40);
    expect(player.volume).toEqual(80);
  });

  it('should reset props on subsequent media changes', async () => {
    await provider.dispatchChange('playbackReady', true);
    player.currentTime = 40;
    player.volume = 80;
    await provider.dispatchLoadStart();
    await provider.dispatchLoadStart();
    await page.waitForChanges();
    expect(player.playbackReady).toBeFalsy();
    expect(player.currentTime).toEqual(0);
    expect(player.volume).toEqual(80);
  });

  it('should autopause player', async () => {
    const playerHtml = '<vime-player><vime-faketube /></vime-player>';
    await buildPage({ html: playerHtml + playerHtml });
    const players = page.body.querySelectorAll('vime-player');
    players[0].paused = false;
    await page.waitForChanges();
    expect(players[0].paused).toBeFalsy();
    players[1].paused = false;
    await page.waitForChanges();
    expect(players[0].paused).toBeTruthy();
    expect(players[1].paused).toBeFalsy();
  });

  it('should not autopause player', async () => {
    const playerHtml = '<vime-player><vime-faketube /></vime-player>';
    await buildPage({ html: playerHtml + playerHtml });
    const players = page.body.querySelectorAll('vime-player');
    players[0].paused = false;
    players[0].autopause = false;
    await page.waitForChanges();
    expect(players[0].paused).toBeFalsy();
    players[1].paused = false;
    await page.waitForChanges();
    expect(players[0].paused).toBeFalsy();
    expect(players[1].paused).toBeFalsy();
  });

  it('should update isTouch when input device changes', async () => {
    provider.dispatchEvent(new Event('touchstart', { bubbles: true }));
    await page.waitForChanges();
    expect(player.isTouch).toBeTruthy();
  });

  it('should bound currentTime between 0 and duration', async () => {
    await provider.dispatchChange('duration', 100);
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    player.currentTime = -1;
    await page.waitForChanges();
    expect(adapter.setCurrentTime).not.toHaveBeenCalled();
    player.currentTime = 101;
    await page.waitForChanges();
    expect(adapter.setCurrentTime).toHaveBeenCalledWith(100);
    player.currentTime = 50;
    await page.waitForChanges();
    expect(adapter.setCurrentTime).toHaveBeenCalledWith(50);
  });

  it('should bound volume between 0 and 100', async () => {
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    player.volume = -1;
    await page.waitForChanges();
    expect(adapter.setVolume).toHaveBeenCalledWith(0);
    player.volume = 101;
    await page.waitForChanges();
    expect(adapter.setVolume).toHaveBeenCalledWith(100);
    player.volume = 40;
    await page.waitForChanges();
    expect(adapter.setVolume).toHaveBeenCalledWith(40);
  });

  it('should default to `en` if language is unknown', async () => {
    const spy = jest.spyOn(console, 'log').mockImplementation();
    player.language = 'cn';
    await page.waitForChanges();
    expect(player.i18n.play).toEqual('Play');
    expect(spy).toHaveBeenCalledWith('[Vime tip]:', expect.stringContaining('invalid `language`'));
    spy.mockRestore();
  });
});

describe('methods', () => {
  it('should return provider when calling getProvider()', async () => {
    // eslint-disable-next-line no-shadow
    const provider = await player.getProvider();
    expect(provider).toEqual(provider);
  });

  it('should call play on adapter when calling play()', async () => {
    expect(adapter.play).not.toHaveBeenCalled();
    await player.play();
    expect(adapter.play).toHaveBeenCalled();
  });

  it('should call pause on adapter when calling pause()', async () => {
    expect(adapter.pause).not.toHaveBeenCalled();
    await player.pause();
    expect(adapter.pause).toHaveBeenCalled();
  });

  it('should call canPlay on adapter when calling canPlay()', async () => {
    adapter.canPlay.mockReturnValueOnce(() => Promise.resolve(true));
    expect(await player.canPlay('')).toBeTruthy();
  });

  it('should throw if calling enterFullscreen() in audio view', async () => {
    await provider.dispatchChange('viewType', ViewType.Audio);
    await page.waitForChanges();
    await expect(player.enterFullscreen()).rejects.toThrow(/audio player view/);
  });

  it('should throw if calling enterFullscreen() and no API is available', async () => {
    await provider.dispatchChange('viewType', ViewType.Video);
    await page.waitForChanges();
    await expect(player.enterFullscreen()).rejects.toThrow(/API is not available/);
  });

  it('should throw if calling enterPiP() in audio view', async () => {
    await provider.dispatchChange('viewType', ViewType.Audio);
    await page.waitForChanges();
    await expect(player.enterPiP()).rejects.toThrow(/audio player view/);
  });

  it('should throw if calling enterPiP() and no API is available', async () => {
    await provider.dispatchChange('viewType', ViewType.Video);
    await page.waitForChanges();
    await expect(player.enterPiP()).rejects.toThrow(/API is not available/);
  });

  it('should extend translations when calling extendLanguage()', async () => {
    await player.extendLanguage('tr', { Play: 'Oyna' });
    await page.waitForChanges();
    expect(player.translations.tr.Play).toEqual('Oyna');
  });
});

describe('events', () => {
  it('should define all events', () => {
    (Object.keys(initialState) as PlayerProp[])
      .map(getEventName)
      .forEach((event) => expect(event).toBeDefined());
  });

  it('should fire change event', async () => {
    const cb = jest.fn();
    page.root!.addEventListener('vPausedChange', cb);
    player.paused = false;
    await page.waitForChanges();
    expect(cb).toHaveBeenCalledTimes(1);
  });

  it('should fire toggle state event', async () => {
    const cb = jest.fn();
    page.root!.addEventListener('vFullscreenChange', cb);
    await provider.dispatchChange('isFullscreenActive', true);
    await page.waitForChanges();
    expect(cb).toHaveBeenCalled();
  });

  it('should fire shortened event', async () => {
    const cb = jest.fn();
    page.root!.addEventListener('vPlaybackStarted', cb);
    await provider.dispatchChange('playbackStarted', true);
    await page.waitForChanges();
    expect(cb).toHaveBeenCalled();
  });

  it('should fire play event', async () => {
    const cb = jest.fn();
    page.root!.addEventListener('vPlay', cb);
    await provider.dispatchChange('paused', false);
    await page.waitForChanges();
    expect(cb).toHaveBeenCalled();
  });

  it('should fire seeked event', async () => {
    const cb = jest.fn();
    page.root!.addEventListener('vSeeked', cb);
    await provider.dispatchChange('seeking', true);
    await page.waitForChanges();
    await provider.dispatchChange('seeking', false);
    await page.waitForChanges();
    expect(cb).toHaveBeenCalled();
  });
});

describe('state changes', () => {
  it('should flush state changes', async () => {
    player.currentTime = 33;
    await provider.dispatchChange('seeking', true);
    await provider.dispatchChange('volume', 66);
    await page.waitForChanges();
    expect(player.currentTime).toEqual(33);
    expect(player.volume).toEqual(66);
    expect(player.seeking).toBeTruthy();
  });

  it('should clear state changes when media changes', async () => {
    const initialCurrentTime = player.currentTime;
    player.currentTime = 33;
    await provider.dispatchLoadStart();
    await provider.dispatchLoadStart();
    await page.waitForChanges();
    expect(player.currentTime).toEqual(initialCurrentTime);
  });

  it('should flush state changes in correct order', async () => {
    await provider.dispatchChange('volume', 33);
    await provider.dispatchChange('volume', 66);
    await provider.dispatchChange('volume', 99);
    await page.waitForChanges();
    expect(player.volume).toEqual(99);
  });
});

describe('lifecycle', () => {
  it('should set attached to true after mounting DOM', async () => {
    expect(player.attached).toBeTruthy();
  });

  it('should set destroyed to true after disconnecting from DOM', async () => {
    player.remove();
    await page.waitForChanges();
    expect(player.attached).toBeFalsy();
  });
});

describe('adapter calls', () => {
  it('should flush playbackReady calls', async () => {
    player.paused = false;
    player.volume = 30;
    player.muted = true;
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    expect(adapter.play).toHaveBeenCalled();
    expect(adapter.setVolume).toHaveBeenCalledWith(30);
    expect(adapter.setMuted).toHaveBeenCalledWith(true);
  });

  it('should queue and flush initial playbackReady calls', async () => {
    await buildPage({
      html: '<vime-player paused="false" muted="true" current-time="50"><vime-faketube /></vime-player>',
    });
    await provider.dispatchChange('duration', 100);
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    expect(adapter.play).toHaveBeenCalled();
    expect(adapter.setCurrentTime).toHaveBeenCalledWith(50);
    expect(adapter.setMuted).toHaveBeenCalledWith(true);
  });

  it('should not reflush playback ready calls', async () => {
    await buildPage({
      html: '<vime-player paused="false" muted="true" current-time="50"><vime-faketube /></vime-player>',
    });
    await provider.dispatchChange('duration', Infinity);
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    expect(adapter.play).toHaveBeenCalledTimes(1);
    expect(adapter.setCurrentTime).toHaveBeenCalledTimes(1);
    expect(adapter.setMuted).toHaveBeenCalledTimes(1);
  });

  it('should not call adapter if change comes from provider', async () => {
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    await provider.dispatchChange('currentTime', 50);
    await page.waitForChanges();
    expect(adapter.setCurrentTime).not.toHaveBeenCalledWith(50);
  });

  it('should call adapter if change comes from user', async () => {
    await provider.dispatchChange('duration', 200);
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    const changes = [];
    for (let i = 0; i < 20; i += 1) {
      // eslint-disable-next-line no-loop-func
      changes.push(async () => {
        provider.dispatchChange('currentTime', Math.floor(Math.random() * 100));
      });
    }
    changes[Math.floor(Math.random() * changes.length)] = async () => { player.currentTime = 33; };
    await Promise.all(changes.map((fn) => fn()));
    await page.waitForChanges();
    expect(adapter.setCurrentTime).toHaveBeenCalledWith(33);
    expect(adapter.setCurrentTime).toHaveBeenCalledTimes(1);
    player.currentTime = 66;
    await page.waitForChanges();
    expect(adapter.setCurrentTime).toHaveBeenCalledWith(66);
    expect(adapter.setCurrentTime).toHaveBeenCalledTimes(2);
  });

  it('should clear queued adapter calls when media changes', async () => {
    player.volume = 30;
    await provider.dispatchLoadStart();
    await provider.dispatchLoadStart();
    await page.waitForChanges();
    await provider.dispatchChange('playbackReady', true);
    await page.waitForChanges();
    expect(adapter.setVolume).not.toHaveBeenCalled();
  });
});

describe('stencil-wormhole', () => {
  it('should pass props through the wormhole to the provider', async () => {
    player.controls = true;
    player.playsinline = true;
    player.muted = true;
    await page.waitForChanges();
    expect(provider.controls).toBeTruthy();
    expect(provider.muted).toBeTruthy();
    expect(provider.playsinline).toBeTruthy();
  });
});
