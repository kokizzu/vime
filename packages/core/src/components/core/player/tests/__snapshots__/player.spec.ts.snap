// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-player aspect-ratio="16:9" buffered="0" current-time="0" duration="-1" id="vime-player-1" language="en" paused="" playback-rate="1" tabindex="0" volume="50">
  <!---->
  <vime-faketube></vime-faketube>
</vime-player>
`;
